### ISSUE TITLE
Safe-cast-and-throw analyzer

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn#23248]
Sometimes writing new code, sometimes refactoring (negate if condition etc), I end up with this:

```cs
if (!(lessDerived is MoreDerived moreDerived)) throw ...
```

I keep forgetting that I really prefer this:

```cs
var moreDerived = lessDerived as MoreDerived ?? throw ...
```

I would love for this to be suggested with a code fix out of the box. If this seems good to you, willing to contribute.

**Comment by sharwell** [comment]
:memo: Right now I'm not too eager to add this due to #22015. I would be curious to see how common the pattern is though.

💭 We'd likely want to handle the related form as well:

```csharp
var moreDerived = lessDerived as MoreDerived;
if (moreDerived == null)
  throw ...;
```

**Comment by jnm2** [comment]
I can only speak for myself but it would gain value for me without waiting for #22015 and it's an increasingly common pattern in the code I work with. Expanding to detect the related form would also be nice.

**Comment by CyrusNajmabadi** [comment]
I think it woudl be a fine fix to add.  It would fit in with a lot of hte other fixes we've added around ??

**Comment by jnm2** [comment]
How much sign-off do I need to start implementing?

**Comment by jmarolf** [comment]
curious what @Pilchie and @jinujoseph's opinions are, but in general we like codefies that promote new language features.

**Comment by jnm2** [comment]
If this would end up exacerbating real end users' debugging issues in the short term I'm happy to wait for the debugging experience to be improved too.

**Comment by CyrusNajmabadi** [comment]
> If this would end up exacerbating real end users' debugging issues in the short term I'm happy to wait for the debugging experience to be improved too.

I don't think you have to wait.  we've already shipped features that push user code toward ```... ?? throw```.  And there's been little (if any) feedback that this was undesirable (at least not for the debugging impact).  So i would not be dissuaded from continuing down this path.

**Comment by jinujoseph** [comment]
Agree with @CyrusNajmabadi  here, don't see a reason to hold the work till the debugging experience improves, If this will provide value add and promote new language features, sounds good to me.
